content/en/blog/_posts/2021-06-08-Kubeadm-Avoid-Default-Address.md

```diff
@@ -0,0 +1,285 @@
+---
+layout: blog
+title: "Avoiding the Default Address when Using kubeadm to Create a Cluster"
+date: YYYY-MM-DD
+slug: avoiding-the-default-address-when-using-kubeadm-to-create-a-cluster
+---
+
+**Authors:** Mike Spreitzer
+
+When using `kubeadm` to create a cluster, the default behavior is to
+take the default address of each node as the address to use in every
+sort of Kubernetes configuration --- and sometimes this is entirely
+wrong.  I am concerning myself here only with IPv4; extension to IPv6
+and dual-stack should be follow the same lines.  The default address
+of a node is the address of the network interface that appears in the
+default route.  In some use cases, this default address is not
+reachable from other nodes and so _must not_ be used in any part of
+Kuberntes configuration.  In a recent adventure of mine, this arose
+because each node is a VirtualBox VM that uses what VirtualBox calls a
+"NAT network adapter" to get connectivity to the outside world.  You
+may have other use cases, perhaps involving containers.
```

Comment:
This'd work better with a bit of a back story / scene setting (IMO). You needed to deploy a hybrid cluster that runs partly on bare metal with interesting constraints. Your cluster will run NetworkPolicy smoketests so doesn't look like a typical CNI. You just can't get enough VirtualBox.

Tell the reader about your recent adventures (in brief) and you will hook them to read on.

Reply:
The real story is really very simple and plain.  I wanted to run a quick experiment using VirtualBox, and I went about it in the most direct way.  If I were willing to put more work into it, I could have created a more elaborate set of VMs, adding a gateway that all the cluster VMs use; that would be a more normal networking topology, and the default address of each cluster node would naturally be distinct.  There are also other ways to not have the problem.  I suspect there are also other ways to have the problem, but can not say I actually ran those scenarios and have not put a lot of thought into building a bullet-proof case for them.  What I can say is that I ran into a problem that arises when you take the most simple and direct approach.

Reply:
I took a stab at this.